xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js

```diff
@@ -55,28 +64,51 @@ window.MSCheckbox = Class.create({
     ];
     this.labels = ['','',''];
 
+    
+    
+    this.button = document.createElement("button");
+    this.button.className = "rights-edit";
+    this.button.onclick = this.createClickHandler(this);
+    
+    var img = document.createElement("img");
+    
+    this.button.appendChild(img);
+    
+    this.label = document.createElement('span');
+    this.label.hidden = true;
+    this.label.appendChild(document.createTextNode(this.labels[this.state]));
+    this.button.appendChild(this.label);
+    
+    $(domNode).appendChild(this.button);
     this.draw(this.state);
-    this.attachEvents();
   },
 
   /**
     * @todo Draw with the current this.state, don't pass as an argument.
     */
   draw: function(state)
   {
-    //remove child nodes
-    while (this.domNode.childNodes.length > 0) {
-      this.domNode.removeChild(this.domNode.firstChild);
-    }
-    //add new image
-    var img = document.createElement('img');
+    //Change display image
+    var img = this.button.firstChild;
     img.src = this.images[state];
-    this.domNode.appendChild(img);
+    
+    //Update the description of the button for accessibility.
+    require(['xwiki-l10n!my-translation-keys'], function(l10n) {
+      var alts = [
+        l10n['inherit'],
+        l10n['allowed'],
+        l10n['denied']
+      ];
+      img.alt = alts[this.state];
+      this.button.title = alts[state];
```

Comment:
I wonder if it is enough to indicate the state. On the one hand, it could be annoying for screen reader users to have more text, on the other hand, how do they know what clicking on the button does? Do you know about best practices for this situation?

Reply:
I think the cleanest thing would be to make the items of the table select inputs as this really models what this is about. The problem is that styling of select elements is extremely limited in HTML and we would be forced to put plain text values in there which would be a significant change and not so nice for the design I fear.

What you suggest sounds good, I think putting the current state first should allow screen reader users to quickly switch to the next item in the table after they heard the first word, but maybe it could be good to confirm this with an actual screen reader user (if this is what they would prefer or if there should, e.g., be some instructions that describe the buttons generically at the start of the table and just the current state on the actual buttons).

Reply:
Btw., I think the respective reference documentation for accessibility of this kind of button is https://www.w3.org/WAI/tutorials/images/functional/#example-5-image-used-in-a-button which says that the `alt` attribute should be used to "convey the purpose of the button".